### PR TITLE
network: TC introduce GRED, Generic Random Early Drop

### PR DESCRIPTION
**Pull request by ssahani**
http://tldp.org/en/Traffic-Control-HOWTO/ar01s06.html

### CODE REVIEW

File: man/systemd.network.xml [hunk 1]
```diff
@@ -2646,7 +2684,7 @@
         </listitem>
       </varlistentry>
     </variablelist>
-  </refsect1>
+ </refsect1>
```
Comment (yuwata) [inline man/systemd.network.xml]: Please drop this change.

File: src/libsystemd/sd-netlink/netlink-types.c [hunk 2]
```diff
@@ -790,10 +790,15 @@ static const NLType rtnl_tca_option_data_tbf_types[] = {
         [TCA_TBF_PBURST]  = { .type = NETLINK_TYPE_U32 },
 };
 
+static const NLType rtnl_tca_option_data_gred_types[] = {
+       [TCA_GRED_DPS]   = { .size = sizeof(struct tc_gred_sopt) },
```
Comment (yuwata) [inline src/libsystemd/sd-netlink/netlink-types.c]: Spurious indentation.

File: src/network/networkd-network-gperf.gperf [hunk 3]
```diff
@@ -281,6 +281,10 @@ FairQueueingControlledDelay.TargetSec,       config_parse_fair_queueing_controll
 FairQueueingControlledDelay.IntervalSec,     config_parse_fair_queueing_controlled_delay_usec,         QDISC_KIND_FQ_CODEL,           0
 FairQueueingControlledDelay.CEThresholdSec,  config_parse_fair_queueing_controlled_delay_usec,         QDISC_KIND_FQ_CODEL,           0
 FairQueueingControlledDelay.ECN,             config_parse_fair_queueing_controlled_delay_bool,         QDISC_KIND_FQ_CODEL,           0
+GRED.Parent,                                 config_parse_qdisc_parent,                                QDISC_KIND_GRED,               0
```
Comment (yuwata) [inline src/network/networkd-network-gperf.gperf:284]: Please rebase and add `GRED.Handle=`

File: src/network/tc/gred.c [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+        struct tc_gred_sopt opt = {};
+        int r;
+
+        assert(link);
+        assert(qdisc);
+        assert(req);
+
+        gred = GRED(qdisc);
+
+
```
Comment (yuwata) [inline src/network/tc/gred.c]: Doubled empty line.

File: src/network/tc/gred.c [hunk 5]
```diff
@@ -0,0 +1,184 @@
...
+        r = sd_netlink_message_open_container_union(req, TCA_OPTIONS, "gred");
+        if (r < 0)
+                return log_link_error_errno(link, r, "Could not open container TCA_OPTIONS: %m");
+
+
```
Comment (yuwata) [inline src/network/tc/gred.c]: Also here.

File: src/network/tc/gred.c [hunk 6]
```diff
@@ -0,0 +1,184 @@
...
+        if (gred->virtual_queues > 0)
+                opt.DPs = gred->virtual_queues;
+
+        if (gred->default_virtual_queue > 0)
+                opt.def_DP = gred->default_virtual_queue;
+
+        if (gred->grio)
+                opt.grio = 1;
```
Comment (yuwata) [inline src/network/tc/gred.c]:
```
if (gred->grio >= 0)
        opt.grio = gred->grio;
```

File: src/network/tc/gred.c [hunk 7]
```diff
@@ -0,0 +1,184 @@
...
+
+        r = sd_netlink_message_append_data(req, TCA_GRED_DPS, &opt, sizeof(struct tc_gred_sopt));
+        if (r < 0)
+                return log_link_error_errno(link, r, "Could not append TCA_GRED_DPS attribute: %m");
+
+
```
Comment (yuwata) [inline src/network/tc/gred.c]: Doubled empty line.

**Comment by ssahani** [comment]
updated thanks for the review @yuwata

File: man/systemd.network.xml [hunk 8]
```diff
@@ -2558,6 +2558,52 @@
...
+        <term><varname>VirtualQueues=</varname></term>
+        <listitem>
+          <para>Specifies the number of virtual queues. Takes a integer in the range 1-16. Defaults to unset and kernel's default is used.</para>
+        </listitem>
+      </varlistentry>
+
+      <varlistentry>
+        <term><varname>DefaultVirtualQueue=</varname></term>
+        <listitem>
+          <para>Specifies the number of default virtual queue. Defaults to unset and kernel's default is used.</para>
```
Comment (yuwata) [inline man/systemd.network.xml]: This must be less than `VirtualQueue=`. Please mention that.

File: man/systemd.network.xml [hunk 9]
```diff
@@ -2558,6 +2558,52 @@
...
+        </listitem>
+      </varlistentry>
+
+      <varlistentry>
+        <term><varname>Grio=</varname></term>
```
Comment (yuwata) [inline man/systemd.network.xml]: What is RIO? Maybe GenericRIO= or GenericRxxxIyyyyOzzzz=.

File: src/network/tc/gred.c [hunk 10]
```diff
@@ -0,0 +1,181 @@
...
+                assert_not_reached("Invalid lvalue.");
+
+        if (isempty(rvalue)) {
+                *p = 0;
+
+                qdisc = NULL;
+                return 0;
+        }
+
+        r = safe_atou32(rvalue, p);
```
Comment (yuwata) [inline src/network/tc/gred.c]: Please do not assign value here. That is,
```
uint32_t v;
r = safe_atou32(rvalue, v);
if (r < 0) {
    log_syntax();
    return 0;
}
if (v > MAX_DPs) {
    log_syntax();
    return 0;
}

*p = v;
```

File: man/systemd.network.xml (hunk 9 above)
Comment (ssahani) [reply man/systemd.network.xml]: I think it's RED In-profile-Out- of-profile.  hence GenericRIO= should be OK.

**Comment by yuwata** [comment]
Included in #14890. Closing.

**Pull request closed without merging**